Add warns, clean up warning handling in tests and improve pytest compatibility

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->


#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

Fixes #15499 

#### Brief description of what is fixed or changed

(Note that this PR description and title is significantly edited since the patch and the rationale changed somewhat while working on this)

This PR is a general clean up of the way that warnings are handled in the sympy tests.

Adds three warning handling utilities to `sympy.utilities.pytest`:
1. `warns` is a context manager similar to `raises`. This is designed to be equivalent to pytest's `warns`. This can be used as `with warns(UserWarning)` to assert in tests that a warning is emitted.
2. `warns_deprecated_sympy` is a shorthand for `warns(SymPyDeprecationWarning)`. The majority of cases where warnings are tested for in the sympy tests are for this case so it warrants a convenience function.
3. `ignore_warnings` is a context manager that can be used to suppress warnings. This is useful for cases where a warning is expected but is not guaranteed to be emitted.

Adds tests for the new warning handling utilities.

Cleans up warning handling code across the sympy tests. This results in changes to many test files.

This fixes all outstanding errors seen when running the non-slow test suite using `pytest` and Python 3.7. Only some of the changes are needed for pytest compatibility: the others are just part of a general cleanup. There are other issues with pytest and Python 2.7 (#15515).

#### Details

Most tests that previously had manual warning handling code are changed to use the three utilities above. In some cases warnings were suppressed that are not emitted. Others had inconsistent or incorrect handling of warnings which is improved by this PR.

Some tests previously modified [global warning filters](https://github.com/sympy/sympy/blob/sympy-1.3/sympy/core/tests/test_args.py#L79) which shouldn't happen because it will affect subsequent tests.

Other test modules [modified global warning handling state at module level](https://github.com/sympy/sympy/blob/sympy-1.3/sympy/geometry/tests/test_line.py#L30). Again this is shouldn't happen because test modules are not guaranteed to be imported in any order and these setting should only be managed in the main test runner code (if at all).

Other tests used `with raises(SymPyDeprecationWarning)` around a block where every line in the block should emit a warning. This means that only the first line of the block was tested for because it raises and exits the block. Using `warns` instead of `raises` ensures that the whole block is tested and that the warning is asserted. To assert that every line raises a warning it is still necessary to use `with warns()` for every line. There is an [example here](https://github.com/sympy/sympy/blob/sympy-1.3/sympy/functions/special/tests/test_delta_functions.py#L67)

In some cases warnings were tested around an overly broad block including several pieces of code [where only one part actually emits a warning](https://github.com/sympy/sympy/blob/sympy-1.3/sympy/multipledispatch/tests/test_core.py#L113). I've rearranged those block so that warnings are tested as precisely as possible.

There are tests in `test_pickling.py` that call `hasattr(w, 'message')` where `w` is an exception or warning instance. On Python 2.7 this triggers a `DeprecationWarning`. I've added a Py2 block of code to suppress these warnings.

The `test_pytest.py` module skipped tests when run under pytest. I've changed it to run the tests under both `bin/test` and `pytest` which enables testing that the utilities provided there are compatible with both.

The `test_source.py` module printed many lines to stdout. I think this is not normally seen when running under `bin/test` because stdout is redirected in some way but this shows up under pytest. I've added a dummy stdout object to capture that in the test.

There were many places that essentially reimplement the `warns` function in different ways. Sometimes these ignore warnings. Sometimes they set warnings to errors and catch them. Sometimes they test for warnings the same way as warns. These are cleaned up to use one or other of the utilities above.

The `import_module` function in `importtools` is changed to emit warnings with `stacklevel=2`. This means that warnings will appear to come from the line that calls `import_module` rather than from within `import_module` itself.

The `bin/test` test runner (`runtest.py`) is changed so that it does not set `WARN_OLD_VERSION` and `WARN_NOT_INSTALLED` to `False`. This enables these warnings throughout the tests. I have instead put warning handlers in the place where the warnings would be raised. This means that the tests are not depending on the altered state of the test runner to work (so that they work when imported normally or when run under pytest).

#### Other comments

In future it should be recommended to use these warning handling utilities rather than reimplementing things in each test file.

One file I didn't resolve is `test_tensor.py` (#15545).

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* utilities
    * Added new utilities for handling warnings in tests. These have been used for a general clean up of warning handling code in the tests.
<!-- END RELEASE NOTES -->

asmeurer commented on sympy/utilities/pytest.py:
```diff
@@ -150,6 +152,36 @@ def func_wrapper():
+            warnings.simplefilter("always")
+            yield
+
+        depwarn = (DeprecationWarning, PendingDeprecationWarning)
```
I think we want to test specifically that SymPyDeprecationWarning is used.

oscarbenjamin commented:
I've just rewritten this to so that it implements/uses `warns` from pytest. Then added a function `sympy_deprecated_call` which can be used like pytest's `deprecated_call` except that it tests whether `SymPyDeprecationWarning was emitted.

asmeurer commented on sympy/utilities/runtests.py:
```diff
@@ -530,11 +530,6 @@ def _test(*paths, **kwargs):
-    # Disable warnings for external modules
-    import sympy.external
-    sympy.external.importtools.WARN_OLD_VERSION = False
-    sympy.external.importtools.WARN_NOT_INSTALLED = False
```
I think these were put in here because if someone has an old version of say matplotlib installed, it should just run the tests skipping the matplotlib tests. 

What is the behavior after this change? Does it just warn or do these warnings produce exceptions?

oscarbenjamin replied:
I think that these settings are there because otherwise various tests would causes warnings to be released. I've moved code to test for those warnings into the tests.

oscarbenjamin replied:
The reason I removed those lines is so that the tests work under `bin.test`, `pytest` *and* when imported and run normally: `import test_module; test_module.test_stuff()`. For those three to be equivalent it's best not to have special settings in the test runner.

asmeurer replied:
Yes but aren't warnings turned into exceptions in the tests (or maybe that's just DeprecationWarning).

oscarbenjamin replied:
Only `DeprecationWarning`s are turned into errors (immediately below the code you're quoting). It looks as if that happens so that `raises` can be used to test for `DepreactionWarning`s.

I could configure `pytest` to also use the same setup as `bin/test` but I think it's better to test with `warns`. Otherwise the tests won't work when run outside the test runner.

oscarbenjamin commented:
I'm going to do a bit more work on this, so it's not yet ready for merging.

bjodah commented:
Thank you for working on this! (It's been on my list for long but never got around to it)

oscarbenjamin committed:
Add warning handlers for use in tests

Part of #15503. This adds warns, ignore_warnings and
warns_deprecated_sympy to be used for warning handling in tests.

oscarbenjamin committed:
Use warning handlers throughout tests

Part of #15503. Uses warns, ignore_warnings, and warns_deprecated_sympy
throughout tests for many modules. Also cleans up warning-handling code
in some tests.

oscarbenjamin committed:
Use dummy stdout to suppress output under pytest

oscarbenjamin committed:
Suppress warnings on Py2 for Warning.message

Part of #15503. Calling e.g. hasattr(w, 'message') for a Warning
instance w triggers DeprecationWarning on Python 2.7. This commit
suppresses 800 warnings seen running the tests under pytest. The special
case can be remoevd when Py2 support is removed.

oscarbenjamin committed:
Add encoding declaration to test_sympy_parser

Part of #15503. The test_sympy_parser module has includes non-ASCII
characters. This is not importable on Py2 without an encoding
declaration. This error does not show up using bin/test because bin/test
uses compile/exec rather than importing the module normally.

oscarbenjamin committed:
Use stacklevel for warnings in import_module

Part of #15503. The warnings emitted from import_module show up in the
importtools module rather than at the import_module call site making it
difficult track down the problem. This commit uses stacklevel=2 so that
the warnings will show up at the call site.

oscarbenjamin committed:
Use warning handlers for tests sympy.external

Part of #15503. Also remove redundant warning handlers from test code.

oscarbenjamin committed:
Suppress warnings from test_latex_deps.py

Part of #15503. Suppresses warnings seen under pytest.

oscarbenjamin committed:
Remove redundant warning-handling code from tests

Part of #15503. Some tests in test_autowrap.py and test_codegen.py have
code to handle/supress warnings although warnings are no longer raised.
This commit removes that redundant code.

oscarbenjamin committed:
Enable warnings for old/missing external modules

Part of #15503. The bin/test test runner disables warnings for external
modules. It is better to suppress these warnings where they are emitted
than globally. This also improves compatiblity with pytest.

oscarbenjamin commented:
I am happy with this PR now.

Note that I've significantly edited the PR and title.

bjodah commented:
Great work, +1 to merge from me.

asmeurer commented:
Can you add something to the docstring of SymPyDeprecationWarning on how to properly test it.

oscarbenjamin committed:
Add docstring for warns_deprecated_sympy

Part of #15503

smichr committed:
Merge branch 'master' into test_warnings
